docs/index.md

```diff
@@ -0,0 +1,130 @@
+* [Twirp golang implementation](https://github.com/twitchtv/twirp)
+* [Twirp wire protocol](https://github.com/twitchtv/twirp/blob/master/PROTOCOL.md)
+* [Protocol Buffers: Intro](https://developers.google.com/protocol-buffers/docs/overview)
+* [Finagle: Intro](https://blog.twitter.com/engineering/en_us/a/2011/finagle-a-protocol-agnostic-rpc-system.html)
+
+## Project setup
+
+To get started with twinagle, you'll need to add a plugin dependency to your project and enable the plugin for your build.
+Add the folowing line to `project/plugins.sbt`:
+
+```scala
+addSbtPlugin("com.soundcloud" % "twinagle-scalapb-plugin" % <version>)
+```
+
+Then, enable the plugin in your build by adding this to `build.sbt`:
+
+```scala
+enablePlugins(Twinagle)
+```
+
+## Defining the API
+
+Twirp APIs are defined in Protobuf files.
+By convention, Twinagle expects them to be under `src/main/protobuf`.
+
+For example, place this under `src/main/protobuf/haberdasher.proto`:
+
+```proto
+syntax = "proto3";
+
+package twitch.twirp.example.haberdasher;
+
+// A Hat is a piece of headwear made by a Haberdasher.
+message Hat {
+  // The size of a hat should always be in inches.
+  int32 size = 1;
+
+  // The color of a hat will never be 'invisible', but other than
+  // that, anything is fair game.
+  string color = 2;
+
+  // The name of a hat is it's type. Like, 'bowler', or something.
+  string name = 3;
+}
+
+// Size is passed when requesting a new hat to be made. It's always measured in
+// inches.
+message Size {
+  int32 inches = 1;
+}
+
+// A Haberdasher makes hats for clients.
+service Haberdasher {
+  // MakeHat produces a hat of mysterious, randomly-selected color!
+  rpc MakeHat(Size) returns (Hat);
+}
+```
+
+## Generating Code
+
+When you compile your project in SBT (e.g. via `sbt compile` or `sbt test`),
+Twinagle will generate code from the API definition.
+
+## Service
+
+The codegen step creates a Service trait that you can extend in order to make a
+proper Finagle service. Example:
+
+```scala
+import twitch.twirp.example.haberdasher.HaberdasherService
+
+class ServiceImpl extends HaberdasherService {
+  override def makeHat(size: Size): Future[Hat] =
+    if (size.inches >= 0) {
+      Future.value(
+        Hat(
+          size = size.inches,
+          color = "brown",
+          name = "bowler"
+        )
+      )
+    } else {
+      Future.exception(TwinagleException(ErrorCode.InvalidArgument, "size must be positive"))
+    }
+}
+
+val httpService: Service[http.Request, http.Response] = HaberdasherService.server(new ServiceImpl())
+```
+
+## Clients
+
+The code generator will create two clients: one that communicates over the wire to
+the matching server using [json](https://developers.google.com/protocol-buffers/docs/proto3#json),
+ and one that uses [binary protobuf](https://developers.google.com/protocol-buffers/docs/encoding).
+
+**Json:**
+
+```scala
+val client = new HaberdasherClientJson(httpService)
```

Comment:
didn't we want to remove the jsonclient? let's do that and not document it?

Reply:
It's part of the Twirp protocol, and also an officially supported proto3 serialization format, so we do want to support it as a general matter in this library.

Reply:
I think we had a discussion about this already and decided to remove it. Happy to discuss this again if needed.

Reply:
The latest twinagle release fixes the issue with the Json client, so we don't need to remove it. Let's mention the binary protobuf client first, and documenting that it should be used unless a human being needs to look at the data on the wire?

Reply:
addressed